### PR TITLE
Split config packages

### LINKED ISSUE #563: Split configuration loading into two package
Currently, dance has structures and functions that load and parse YAML configuration files and transform them into a form used by other packages all in one package. Some `struct`s have two representations (`Results` and `ResultsYAML`), while others have only one used in both contexts (`Stats`). That makes it harder to introduce new features (#463) and already lead to one abandoned PR (#327).

We should split this code into two packages: `internal/config` and `internal/configload`. The first one should be used by other packages and "business logic", while the second should export a single `Load` function that would return `internal/config.Config` `struct`.

Do not do other features or refactorings in this issue.

### PR DESCRIPTION
**Pull request by b1ron**
Closes #563.

### CODE REVIEW

File: internal/config/config.go [hunk 1]
```diff
@@ -23,30 +25,84 @@ import (
...
+	RunnerTypeYCSB RunnerType = "ycsb"
+)
+
+// RunnerType represents the type of test runner used in the configuration.
+type RunnerType string
+
+// Stats represent the expected/actual amount of
+// failed, skipped and passed tests.
+type Stats struct {
+	UnexpectedRest int `yaml:"unexpected_rest"`
```
Comment (AlekSi) [inline internal/config/config.go:48]: This package should not have structs with YAML tags

File: internal/configload/configload.go [hunk 2]
```diff
@@ -12,113 +12,131 @@
...
 
-// ConfigYAML is a yaml tests representation of the Config struct.
-//
-// It is used only to fetch data from file. To get any of
-// the dance configuration data it should be converted to
-// Config struct with Convert() function.
+// ConfigYAML represents the YAML-based configuration for the testing framework.
 //
 //nolint:govet // we don't care about alignment there
 type ConfigYAML struct {
```
Comment (AlekSi) [inline internal/configload/configload.go:31]: Those types should not be exported

File: internal/configload/configload.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+	"golang.org/x/exp/maps"
+	"gopkg.in/yaml.v3"
+
+	"github.com/FerretDB/dance/internal/config"
+)
+
+// configYAML represents the YAML-based configuration for the testing framework.
+//
+//nolint:govet // we don't care about alignment there
+type configYAML struct {
```
Comment (b1ron) [inline internal/configload/configload.go:31]: I can remove the YAML suffix if needed.

File: internal/configload/configload.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+		if !validStatus(r.Default) {
+			return fmt.Errorf("invalid default result: %q", *origDefault)
+		}
+
+		if *commonDefault != "" && r.Default != "" {
+			return errors.New("default value cannot be set in common, when it's set in database")
+		}
+
+		// XXX this doesn't work
+		r.Default = *commonDefault
```
Comment (b1ron) [inline internal/configload/configload.go:275]: We can't set the default value here, so we have to set it [here](https://github.com/FerretDB/dance/pull/570/files#diff-54c7c1af5fa8d5db4dc49f0e8e80e93ba2b1183ba4d5c9e2e5729e6deae6a3cdR181) instead which makes this method incomplete.

File: internal/configload/configload.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+		origDefault = commonDefault
```
Comment (b1ron) [inline internal/configload/configload.go:280]: We can't set the default value here, so we have to set it [here](https://github.com/FerretDB/dance/pull/570/files#diff-54c7c1af5fa8d5db4dc49f0e8e80e93ba2b1183ba4d5c9e2e5729e6deae6a3cdR181) instead which makes this method incomplete.

File: internal/config/config_test.go [hunk 6]
```diff
@@ -12,178 +12,217 @@
...
 				},
-				MongoDB: &TestsConfig{
+				MongoDB: &TestConfig{
 					Default: "pass",
 				},
 			},
 		},
-		"FillAndValidateDefaultDuplicate": {
+		// XXX fails because this error is now returned in fillAndValidate
+		"MergeTestConfigsDefaultDuplicate": {
```
Comment (b1ron) [inline internal/config/config_test.go:187]: I need to write a new test for this.

File: internal/configload/configload.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+	validStatus := func(status ic.Status) bool {
+		s := ic.Status(strings.ToLower(string(status)))
+		_, ok := ic.KnownStatuses[s]
+
+		return ok
+	}
+
+	if c.Results.Common.Default == nil {
+		c.Results.Common.Default = new(ic.Status)
```
Comment (b1ron) [inline internal/configload/configload.go:248]: `ic.Status` is not a composite type so I cannot allocate it using that syntax, right?

File: internal/config/config.go [hunk 8]
```diff
@@ -23,42 +25,195 @@ import (
...
+	Runner RunnerType
+	Dir    string
+	// Args contains additional arguments for the test runner.
+	Args    []string
+	Results Results
+}
+
+// Results stores the expected test results for different databases.
+type Results struct {
+	Common   *TestConfig
```
Comment (AlekSi) [inline internal/config/config.go:72]: Let's remove that field. The idea that `configload` will do all needed transformations, merges, etc, and then the rest of the code could work with a simple `Config` type by using `Config.FerretDB` or `Config.For("ferretdb")`

Comment (b1ron) [reply internal/config/config.go]: Do you mean here

File: internal/config/config.go [hunk 9]
```diff
@@ -23,42 +24,137 @@ import (
...
+
+	// RunnerTypeYCSB indicates a YCSB test runner.
+	RunnerTypeYCSB RunnerType = "ycsb"
+)
+
+// RunnerType represents the type of test runner used in the configuration.
+type RunnerType string
+
+// Stats represent the expected/actual amount of
+// failed, skipped and passed tests.
```
Comment (AlekSi) [inline internal/config/config.go:45]: FWIW, I think we should follow the same order of fields as in this comment: failed, skipped, passed. Think of it like a traffic lights: 🔴, 🟡, 🟢.

File: internal/config/config.go [hunk 10]
```diff
@@ -279,3 +381,13 @@ func (tc *TestsConfig) toMap() (map[string]status, error) {
 
 	return res, nil
 }
+
+// CheckDuplicates checks for duplicate test names or prefixes within a *TestConfig instance.
+// It returns an error if any duplicates are found.
+func (tc *TestConfig) CheckDuplicates() error {
```
Comment (AlekSi) [inline internal/config/config.go:387]: That should be in the `configload` package. Once the config is loaded, it is valid, does not contain duplicates, etc

File: internal/configload/configload.go (hunk 7 above)
Comment (b1ron) [reply internal/configload/configload.go:252]: @AlekSi is there a directive I can use to ignore this in the linter?

**Review by AlekSi — COMMENTED** [review]
CI fails. Please run linters locally

**Comment by b1ron** [comment]
> CI fails. Please run linters locally

I did and fixed what could be fixed. The other failure is related to a syntax error, see [here](https://github.com/FerretDB/dance/actions/runs/6115959796/job/16600415356?pr=570#step:4:104). I mentioned it in [this](https://github.com/FerretDB/dance/pull/570#discussion_r1315143899) comment as well.

File: internal/configload/configload.go [hunk 11]
```diff
@@ -249,8 +250,7 @@ func (c *config) fillAndValidate() error {
 
 	// initialize default common field if it's nil
 	if c.Results.Common.Default == nil {
-		c.Results.Common.Default = new(ic.Status)
-		*c.Results.Common.Default = ic.Pass
+		c.Results.Common.Default = pointer.To(ic.Pass)
```
Comment (b1ron) [inline internal/configload/configload.go:253]: Oh wow!

**Review by AlekSi — APPROVED** [review]

**Pull request merged**